Tools/cases_generator/tier2_abstract_generator.py

```diff
@@ -0,0 +1,457 @@
+            if var.name in UNUSED:
+                continue
+            out.emit(f"{var.name} = get_const({mangled_var.name});\n")
+        emit_tokens(out, uop, stack, None, TIER2_REPLACEMENT_FUNCTIONS)
+        out.emit("\n")
+        # Guard elimination
+        out.emit('DPRINTF(3, "const eliminated guard\\n");\n')
+        out.emit("new_inst.opcode = _NOP;")
+        out.emit("break;\n")
+        out.emit("}\n")
+
+    # Does the input specify typed inputs?
+    if not any(output_var.type_prop for output_var in mangled_uop.stack.outputs):
+        return
+    # If the input types already match, eliminate the guard
+    # Read the cache information to check the auxiliary type information
+    predicates = []
+    propagates = []
+
+    assert len(mangled_uop.stack.outputs) == len(
+        mangled_uop.stack.inputs
+    ), "guards must have same number of args"
+    assert [
+        output == input_
+        for output, input_ in zip(mangled_uop.stack.outputs, mangled_uop.stack.inputs)
+    ], "guards must forward their stack values"
+    for output_var in mangled_uop.stack.outputs:
+        if output_var.name in UNUSED:
+            continue
+        if (typ := output_var.type_prop) is not None:
+            typname, aux = typ
+            aux = "0" if aux is None else aux
+            # Check that the input type information match (including auxiliary info)
+            predicates.append(
+                f"sym_matches_type((_Py_UOpsSymbolicValue *){output_var.name}, {typname}, (uint32_t){aux})"
+            )
+            # Propagate mode - set the types
+            propagates.append(
+                f"sym_set_type((_Py_UOpsSymbolicValue *){output_var.name}, {typname}, (uint32_t){aux})"
+            )
+
+    out.emit("// Type guard elimination\n")
+    out.emit(f"if ({' && '.join(predicates)}){{\n")
+    out.emit('DPRINTF(2, "type propagation eliminated guard\\n");\n')
+    out.emit("new_inst.opcode = _NOP;")
+    out.emit("break;\n")
+    out.emit("}\n")
+    # Else we need the guard
+    out.emit("else {\n")
+    out.emit("// Type propagation\n")
+    for prop in propagates:
+        out.emit(f"{prop};\n")
+    out.emit("}\n")
+
+
+def write_abstract_uop(mangled_uop: Uop, uop: Uop, out: CWriter, stack: Stack) -> None:
+    try:
+        out.start_line()
+        is_impure = not mangled_uop.properties.pure and not mangled_uop.properties.guard
+        # These types of guards do not need the stack at all.
+        if not (
+            mangled_uop.properties.guard
+            and mangled_uop.name in NO_CONST_OR_TYPE_EVALUATE
+        ):
+            for var in reversed(mangled_uop.stack.inputs):
+                old_var_name = var.name
+                # code smell, but basically impure ops don't use any of their inputs
+                if is_impure:
+                    var.name = "unused"
+                out.emit(stack.pop(var))
+                var.name = old_var_name
```

Comment:
I think you can rewrite this (and the one below) cleaner:
```suggestion
                text = stack.pop(var)
                if not is_impure:
                    out.emit(text)
```
(Maybe invent a better name than `text`?)